site/app/libraries/database/DatabaseQueries.php

```diff
@@ -722,6 +722,50 @@ public function getCategoriesIdForThread($thread_id) {
         return $categories_list;
     }
 
+    public function toggleLikes($post_id, $thread_id, $current_user){
+        try{
+            $this->course_db->query("SELECT * FROM forum_upducks WHERE post_id = ? AND user_id = ? AND thread_id = ?", [$post_id, $current_user, $thread_id]);
+
+            if (isset($this->course_db->rows()[0])) {
+                // If a row with the same values exists, delete it
+                $this->course_db->query("DELETE FROM forum_upducks WHERE post_id = ? AND user_id = ? AND thread_id = ?", [$post_id, $current_user, $thread_id]);
+                $result = 'unlike';
+            } 
+            else {
+                // If no row with the same values exists, insert the new row
+                $this->course_db->query("INSERT INTO forum_upducks (post_id, user_id, thread_id) VALUES (?, ?, ?)",[$post_id, $current_user, $thread_id]);
+                $result = 'like';
+            }
+            return $result;
+        }   
+        catch (DatabaseException $dbException) {
+            if ($this->course_db->inTransaction()) {
+                $this->course_db->rollback();
+            }
+            return false;
+        }
+    }
+
+    public function getUpduckInfoForPosts(array $post_ids): array {
+        $resultDict = [];
+        foreach($post_ids as $post_id) {
+            $this->course_db->query('SELECT COUNT(*) AS cnt FROM forum_upducks WHERE post_id = ?', [$post_id]);
```

Comment:
The idea behind moving the queries out of createPost is so that you won't run queries within a loop. Queries are slow and it's a lot faster to iterate over a large result from a query than to make a bunch of smaller queries.